File: Day188_2025/Q188_Leetcode.java
import java.util.*;

class Q188_Leetcode {
    public int maxEvents(int[][] events) {
        Arrays.sort(events, (a, b) -> a[0] - b[0]);  // sort by start day
        PriorityQueue<Integer> pq = new PriorityQueue<>(); // min-heap of end days

        int i = 0, n = events.length, res = 0;
        int day = 0;
        
        // Find the last day any event ends
        int maxDay = 0;
        for (int[] event : events) {
            maxDay = Math.max(maxDay, event[1]);
        }

        for (day = 1; day <= maxDay; day++) {
            // Push events starting today
            while (i < n && events[i][0] == day) {
                pq.offer(events[i][1]);
                i++;
            }

            // Remove events that have already expired
            while (!pq.isEmpty() && pq.peek() < day) {
                pq.poll();
            }

            // Attend one event today
            if (!pq.isEmpty()) {
                pq.poll();
                res++;
            }
        }

        return res;
    }
}

File: Day188_2025/Q188_Leetcode.py
import heapq
from typing import List

class Solution:
    def maxEvents(self, events: List[List[int]]) -> int:
        events.sort()
        n = len(events)
        max_day = max(e[1] for e in events)
        pq = []
        res = 0
        i = 0

        for day in range(1, max_day + 1):
            # Add events that start today
            while i < n and events[i][0] == day:
                heapq.heappush(pq, events[i][1])
                i += 1

            # Remove past events
            while pq and pq[0] < day:
                heapq.heappop(pq)

            # Attend one event
            if pq:
                heapq.heappop(pq)
                res += 1

        return res
